Add relevantForJS label frame support for js implementation filter

canaltinova opened the pull request:
Well, I was looking from another perspective for the whole thing and after talking with Greg, I understood that it's actually pretty easy to implement because we have one mechanism that all transforms use(`FUNC_MATCHES`).
For this PR, the remaining part is handling URL serialized call tree transforms with URL upgraders.

canaltinova commented:
@gregtatum It's ready for the review now. Added the upgraders/tests and fixed all the failures. Could you review it?
@julienw also adding you as a reviewer for one commit that Greg wrote. It would be great if you could review it.

gregtatum commented on src/components/app/UrlManager.js:
```diff
@@ -38,7 +51,37 @@ type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;
+      // Just silently finish the url setup and return to home.
+      urlSetupDone();
+    } else {
+      throw new Error('This if branch should not be reached.');
```
Maybe explain why it should not be reached, e.g. "An an unhandled case was reached during the initial processing of URLs."

gregtatum commented on src/components/app/UrlManager.js:
```diff
@@ -105,24 +145,33 @@ class UrlManager extends React.PureComponent<Props> {
+    const { urlSetupPhase, children } = this.props;
+    switch (urlSetupPhase) {
+      case 'initial-load':
+        return <div className="processingUrl" />;
```
Is this div ever actually presented to the user? It might be better to return null instead, since the div with a className implies some kind of presentation to the user. I'm guessing this pathway is completely synchronous.

gregtatum commented on src/reducers/profile-view.js:
```diff
@@ -29,7 +29,7 @@ import type {
 const profile: Reducer<Profile | null> = (state = null, action) => {
   switch (action.type) {
-    case 'VIEW_PROFILE':
+    case 'PROFILE_LOADED':
```
Do we actually need to store the profile in the store here? Could it be loaded via thunk action, and only added to the store via `VIEW_PROFILE`? I think this may make the action and state changes more atomic, and reduce the multiple dispatches. I may be missing some necessary step here, but I think this may be possible.

canaltinova replied:
I'm not sure if we really want to do that. Since `VIEW_PROFILE` is now more like a setter for the profile view, and `PROFILE_LOADED` is more like a setter of the profile. We don't set the profile to store in `VIEW_PROFILE` anymore. We can return the profile in `loadProfile` thunk action and then postpone the profile storing until `VIEW_PROFILE` but then we'll have to also return everything else we store in `PROFILE_LOADED` too(pathInZipFile, implementationFilter, transformStacks). I think storing them at that moment is more convenient. What do you think?

julienw commented on src/app-logic/url-handling.js:
```diff
@@ -552,3 +552,64 @@ if (Object.keys(_upgraders).length - 1 !== CURRENT_URL_VERSION) {
+      stackMatchesCallNodePath.add(stackIndex);
+      stackIndexDepth.set(stackIndex, depth);
+    } else {
+      // Any non-JS stack potentially matches, add it here.
```
I'd expand a bit more on this comment.
Maybe something like:
```js
// Any non-JS stack potentially matches, because they're skipped in the JS call node path.
// Let's add it here using the previous depth.
```

julienw commented on src/app-logic/url-handling.js:
```diff
@@ -552,3 +618,64 @@ if (Object.keys(_upgraders).length - 1 !== CURRENT_URL_VERSION) {
   `);
 }
+
+function getStackIndexFromOldJSCallNodePath(
```
Also I'd use the version numbers in the names, for example `getStackIndexFromVersion3JSCallNodePath` here, and `getVersion4JSCallNodePathFromStackIndex` for the other.

julienw commented on src/app-logic/url-handling.js:
```diff
@@ -536,14 +552,64 @@ const _upgraders = {
+        transform.implementation !== 'js' ||
+        !transform.callNodePath
+      ) {
+        // Only transfroms with JS implementation filters that have callNodePaths
```
```suggestion
        // Only transforms with JS implementation filters that have callNodePaths
```

canaltinova commented:
@gregtatum Addressed most of the review comments and added some message to the rest. Added the changes as additional commits or fixup commits instead of rebasing them to make it easier to review. Will rebase those before landing. Could you take a look at it again?

mstange commented on src/app-logic/url-handling.js:
```diff
@@ -552,3 +552,64 @@ if (Object.keys(_upgraders).length - 1 !== CURRENT_URL_VERSION) {
+    const funcIndex = frameTable.func[frameIndex];
+    const isJS = funcTable.isJS[funcIndex];
+    const doesPrefixMatchCallNodePath =
+      prefix === null || stackMatchesCallNodePath.has(prefix);
```
Yes, this is always the case, and we rely on it everywhere. If the sortedness assumption is not documented at the definition of the stack table (or its flow type definition?) then we should fix the comments.

gregtatum commented on src/app-logic/url-handling.js:
```diff
@@ -653,18 +659,19 @@ function getStackIndexFromOldJSCallNodePath(
 
-function getJSCallNodePathFromStackIndex(
+// Constructs the new JS CallNodePath from given stackIndex and returns it.
+// This should only be used for the URL upgrader.
```
Thanks, these are much scarier sounding functions now :)

julienw commented on src/components/app/UrlManager.js:
```diff
@@ -77,7 +77,9 @@ class UrlManager extends React.PureComponent<Props> {
     } else {
-      throw new Error('This if branch should not be reached.');
+      throw new Error(
+        'An an unhandled case was reached during the initial processing of URLs'
```
Typo: you have "an" twice at the start of the sentence :)

canaltinova replied:
ahh, thanks! fixed it :)

canaltinova commented on src/components/app/UrlManager.js:
```diff
@@ -36,9 +49,67 @@ type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;
 /**
  * This component manages the interaction with the window.history browser API.
+ *
+ * There are two different profile loading paths currently:
```
Added some documentation about the different profile loading paths. Let me know what you think.

canaltinova commented:
Added a comment that explains those two dispatches in `loadProfile` function.

gregtatum commented on src/components/app/UrlManager.js:
```diff
@@ -36,9 +49,67 @@ type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;
+ * There are two different profile loading paths currently:
+ * 1. Initial URL processing:
+ *    This happens when user loads a URL with a profile hash or URL.
+ *    In the first load, have the URL but we don't have the `UrlState` to get the
```
```suggestion
 *    In the first load, we have the URL but we don't have the `UrlState` to get the
```

gregtatum commented on src/components/app/UrlManager.js:
```diff
@@ -36,9 +49,67 @@ type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;
+ *       it, or retrieve from Firefox. Set the `urlSetupPhase` as 'loading-profile'.
+ *    4. Upgrade the URL with downloaded profile data and setup initial `UrlState`.
+ *    5. Finalize the profile view with the `UrlState` information. (we couldn't
+ *       finalize the prifile view immediately after the download because we need
```
```suggestion
 *       finalize the profile view immediately after the download because we need
```

gregtatum commented on src/components/app/UrlManager.js:
```diff
@@ -36,9 +49,67 @@ type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;
+ *       `UrlState` for this step.)
+ *    6. Set the `urlSetupPhase` as 'done' and display the profile view.
+ *
+ * 2. State changes after some user interractions(e.g. drag and drop):
```
```suggestion
 * 2. State changes after some user interactions (e.g. drag and drop):
```

canaltinova commented:
Filed #2109 for that. Thanks for pointing out the typos! :)
